sycl/source/detail/graph_impl.cpp

```diff
@@ -1675,20 +1650,133 @@ void exec_graph_impl::updateImpl(std::shared_ptr<node_impl> Node) {
   UpdateDesc.pNewLocalWorkSize = LocalSize;
   UpdateDesc.newWorkDim = NDRDesc.Dims;
 
+  // Query the ID cache to find the equivalent exec node for the node passed to
+  // this function.
+  // TODO: Handle subgraphs or any other cases where multiple nodes may be
+  // associated with a single key, once those node types are supported for
+  // update.
+  auto ExecNode = MIDCache.find(Node->MID);
+  assert(ExecNode != MIDCache.end() && "Node ID was not found in ID cache");
+
   ur_exp_command_buffer_command_handle_t Command =
       MCommandMap[ExecNode->second];
-  ur_result_t Res = Adapter->call_nocheck<
-      sycl::detail::UrApiKind::urCommandBufferUpdateKernelLaunchExp>(
-      Command, &UpdateDesc);
+  UpdateDesc.hCommand = Command;
+
+  // Update ExecNode with new values from Node, in case we ever need to
+  // rebuild the command buffers
+  ExecNode->second->updateFromOtherNode(Node);
+}
+
+std::map<std::shared_ptr<partition>, std::vector<std::shared_ptr<node_impl>>>
+exec_graph_impl::getPartitionForNodes(
+    const std::vector<std::shared_ptr<node_impl>> &Nodes) {
+  // Iterate over each partition in the executable graph, and find the nodes
+  // in "Nodes" that also exist in the partition.
+  std::map<std::shared_ptr<partition>, std::vector<std::shared_ptr<node_impl>>>
+      PartitionedNodes;
+  for (const auto &Partition : MPartitions) {
+    std::vector<std::shared_ptr<node_impl>> NodesForPartition;
+    const auto PartitionBegin = Partition->MSchedule.begin();
+    const auto PartitionEnd = Partition->MSchedule.end();
+    for (auto &Node : Nodes) {
+      auto ExecNode = MIDCache.find(Node->MID);
+      assert(ExecNode != MIDCache.end() && "Node ID was not found in ID cache");
+
+      if (std::find_if(PartitionBegin, PartitionEnd,
+                       [ExecNode](const auto &PartitionNode) {
+                         return PartitionNode->MID == ExecNode->second->MID;
+                       }) != PartitionEnd) {
+        NodesForPartition.push_back(Node);
+      }
+    }
+    if (!NodesForPartition.empty()) {
+      PartitionedNodes.insert({Partition, NodesForPartition});
+    }
+  }
```

Comment:
This seems potentially quite expensive to be doing during the update call (having to search through the partition nodes repeatedly). Could we cache the partition that a node belongs to (since this will be known when we do the partitioning during finalize), such that we can easily skip nodes that are not part of this partition or restructure it such that we only loop over `Nodes` and just add each one to the appropriate partition map. 

In the latter case I guess we'd have to insert empty vectors before the loop for each partition and then retrieve them inside the loop over the nodes, but it feels like that should probably still be faster, not sure.

Reply:
I've realized that there is already a `std::unordered_map<std::shared_ptr<node_impl>, int> MPartitionNodes` member in the executable graph object, so I've managed to change this so that we only iterate over the list of nodes once and use the partition index from that map to assign it to one of the returned map lists.

Pushed that as a new change to make it easier to review

Reply:
Change looks good to me!